### PR TITLE
Update openenv examples to use `environment_factory`

### PR DESCRIPTION
**Pull request by sergiopaniego**
# What does this PR do?

TODO:

- [x] Migrate notebooks
- [x] Update TRL-OpenEnv guide
- [x] Add multi-env example

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

---

> [!NOTE]
> **Low Risk**
> Documentation-only changes, but they significantly rewrite OpenEnv training guidance and example code; risk is mainly confusing users if the new `environment_factory` patterns or links are incorrect.
> 
> **Overview**
> **Migrates OpenEnv examples to `environment_factory`.** The OpenEnv integration guide is heavily rewritten to center on `GRPOTrainer(environment_factory=...)`, including updated install instructions, environment-class requirements, reward functions consuming the new `environments` argument, and guidance on multi-turn limits and server concurrency.
> 
> **Reorganizes OpenEnv docs/examples.** Adds `OpenEnv` to the main docs toctree under Integrations, restructures `example_overview.md` and `examples/notebooks/README.md` to introduce dedicated *OpenEnv Notebooks*/*OpenEnv Scripts* sections, and updates the Sudoku GRPO notebook to remove the custom rollout pipeline in favor of an `environment_factory`-based `SudokuEnv` plus environment-backed reward functions (and includes a multi-environment training example in the OpenEnv guide).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit fee1876f2a350447973c43c5906a8637fd81869b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by sergiopaniego** [comment]
**PR: Update OpenEnv examples and docs to `environment_factory`** generated by Claude

## Summary

- Migrated all OpenEnv example scripts and notebooks from `rollout_func` to `environment_factory`: sudoku, catch, browsergym_llm (scripts) and wordle, sudoku (notebooks)
- New `multi_env.py` example script demonstrating multi-environment GRPO training (Wordle + Catch in the same run) with per-environment reward functions returning `None` for non-applicable environments
- Rewrote `docs/source/openenv.md` entirely around `environment_factory`: environment class pattern, tool discovery via introspection, Echo and Wordle tutorials, multi-environment section, server concurrency configuration, and a migration guide from `rollout_func`
- Reorganized `example_overview.md` and `examples/notebooks/README.md` with dedicated OpenEnv subsections
- Updated all OpenEnv documentation URLs to current paths (`from_hub` removed, `.html` extensions)

## Trainer fixes (`grpo_trainer.py`)

- Fixed `self.llm` → `self.vllm_generation.llm` for colocate mode in the tool-calling loop
- Added server mode branch for `max_model_len` (was raising `NotImplementedError`)
- Added done-check (`_done`/`done`) to stop calling tools for finished environments
- Added size alignment between `tool_mask`/`logprobs`/`completion_ids` after the tool loop (re-tokenization can cause off-by-one mismatches)
- Added truncation to `max_completion_length` to prevent size mismatch errors in long multi-turn episodes
- Fixed `nan` metrics crashing JSON serialization in loggers (convert to `None`). This happens in multi-environment training where per-environment reward functions return `None` for non-applicable samples, resulting in `nan` averages

## vLLM fix (`vllm_generation.py`)

- Added `tools` parameter to `generate()` so colocate mode stops generation at `</tool_call>`, allowing the tool-calling loop to execute tools between turns
- Passed `self.tools` from the trainer to enable this

## Not migrated

- `browsergym.py`: requires multimodal tool responses (screenshots) not yet supported in `_tool_call_loop`. To migrate, `_tool_call_loop` needs to: (1) allow tool methods to return structured content (`[{"type": "image", ...}]`), (2) detect multimodal returns and build proper content blocks instead of `str()`, and (3) run image processing (`pixel_values`, `image_grid_thw`) per tool-calling iteration, not just once at generation start
- `nemo_gym`: fundamentally incompatible, external agent servers own generation
- `browsergym notebook`: blocked by Playwright/greenlet server issue and model compatibility (FunctionGemma, not Qwen3). To migrate: (1) upgrade openenv-core to async v0.2.2+ so Playwright runs in a dedicated thread pool instead of conflicting with FastAPI's event loop, and (2) investigate Qwen3 compatibility or adapt the notebook to work with FunctionGemma's tool-calling format

## Not in this PR

- Upgrade to openenv-core async v0.2.2+ (server + client `.sync()`)
- Update HF Space URLs to official openenv org (some scripts still point to personal spaces)

## Required HF Space configuration

`environment_factory` creates N concurrent environment instances (one per generation), each opening a WebSocket to the server. By default, OpenEnv servers only allow 1 concurrent session, so the Spaces need to be configured for concurrency:

1. In `server/environment.py`, declare concurrent session support:
```python
SUPPORTS_CONCURRENT_SESSIONS: bool = True
```

2. In `server/app.py`, set the concurrency limit:
```python
app = create_app(
    create_my_environment,
    MyAction,
    MyObservation,
    max_concurrent_envs=64,  # must be >= generation_batch_size
)
```

`max_concurrent_envs` should be >= `per_device_train_batch_size * gradient_accumulation_steps`. For example, with `gradient_accumulation_steps=64` and batch size 1, you need at least 64 concurrent sessions.

This applies to all Spaces used with `environment_factory`: echo, wordle, catch, sudoku, browsergym_llm, carla.

File: trl/generation/vllm_generation.py [hunk 1]
```diff
@@ -650,6 +651,12 @@ def generate(
             }
             generation_kwargs.update(self.generation_kwargs)
 
+            # When tools are provided, stop generation after the first tool call so the tool-calling
+            # loop can execute the tool and feed the result back before the next generation.
+            if tools and "stop" not in generation_kwargs:
+                generation_kwargs["stop"] = ["</tool_call>"]
```
Comment (qgallouedec) [inline trl/generation/vllm_generation.py]: why do you need this? the model should output an EOS right after `</tool_call>` (btw we have no guarantee that all models use the str `</tool_call>` so we can't use it here

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1494,11 +1498,51 @@ async def _run_async_tools(async_coros):
...
-            tool_calls = [tool_call for tool_call in tool_calls if tool_call]
+            filtered_idxs = []
+            filtered_tool_calls = []
+            for idx, tool_call in zip(idxs_with_tool, tool_calls, strict=True):
+                if not tool_call:
+                    continue
+                # If the environment signals it's done, stop calling tools for it
+                env = self.environments[idx]
+                if getattr(env, "_done", False) or getattr(env, "done", False):
+                    continue
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: This feels like extending the assumptions we make about what an environment is. To ensure max compatibility, we should aim for a definition of an environment as minimal as possible, so unless this is strictly necessary I’d prefer not to introduce additional structure.

In particular, I don’t think we need special handling for `done`. The environment can simply include that signal in the observation it returns, and the model should learn not to query the environment again after that point.

If respecting `done` is important, the environment can raise an exception when it is called after termination. That makes the constraint explicit and helps the model learn that calling the environment again in that state is invalid.

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1494,11 +1498,51 @@ async def _run_async_tools(async_coros):
...
+                elif lp_len < comp_len:
+                    logprobs[i] = logprobs[i] + [0.0] * (comp_len - lp_len)
+
+        # Truncate to max_completion_length to prevent size mismatches downstream
+        for i in range(len(completion_ids)):
+            if len(completion_ids[i]) > self.max_completion_length:
+                completion_ids[i] = completion_ids[i][: self.max_completion_length]
+                tool_mask[i] = tool_mask[i][: self.max_completion_length]
+                if logprobs is not None:
+                    logprobs[i] = logprobs[i][: self.max_completion_length]
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: everything here can be removed, we're addressing the issue in https://github.com/huggingface/trl/issues/5224

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -2292,7 +2336,10 @@ def prediction_step(self, model, inputs, prediction_loss_only, ignore_keys: list
 
     def log(self, logs: dict[str, float], start_time: float | None = None) -> None:
         mode = "train" if self.model.training else "eval"
-        metrics = {key: sum(val) / len(val) for key, val in self._metrics[mode].items()}  # average the metrics
+        metrics = {}
+        for key, val in self._metrics[mode].items():
+            avg = sum(val) / len(val)
+            metrics[key] = None if math.isnan(avg) else avg
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: why you need this?

File: trl/trainer/grpo_trainer.py (hunk 3 above)
Comment (sergiopaniego) [reply trl/trainer/grpo_trainer.py]: removed. In this case, we'd need to merge this PR after #5242

File: trl/trainer/grpo_trainer.py (hunk 4 above)
Comment (sergiopaniego) [reply trl/trainer/grpo_trainer.py]: It handles multi-env training where reward functions return None for non-applicable samples (when training on "the other" env basically). For example, when training on wordle + catch simultaneously, wordle_reward returns None for all catch samples and vice versa. If an entire batch happens to contain only one environment's samples, the per-reward metrics average to NaN, which crashes JSON loggers in trackio. It converts NaN to None for clean serialization.

**Comment by qgallouedec** [comment]
@codex review

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -2457,7 +2460,13 @@ def prediction_step(self, model, inputs, prediction_loss_only, ignore_keys: list
...
     def log(self, logs: dict[str, float], start_time: float | None = None) -> None:
         mode = "train" if self.model.training else "eval"
-        metrics = {key: sum(val) / len(val) for key, val in self._metrics[mode].items()}  # average the metrics
+        # Average the metrics
+        metrics = {}
+        for key, val in self._metrics[mode].items():
+            avg = sum(val) / len(val)
+            # If a reward function returns None for all samples in a batch, its metric is NaN. Convert to None
+            # for clean serialization (e.g. JSON loggers crash on float NaN).
+            metrics[key] = None if math.isnan(avg) else avg
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: Yes @sergiopaniego Please address this

File: examples/scripts/openenv/browsergym_llm.py [hunk 6]
```diff
@@ -406,11 +268,141 @@ def reward_completion(completions: list[str], **kwargs) -> list[float]:
...
+            """Fill an input field with text.
+
+            Args:
+                bid: The BrowserGym ID of the input field.
+                text: The text to type into the field.
+
+            Returns:
+                The updated page observation.
+            """
+            return self._do_action(f"fill('{bid}', '{text}')")
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/scripts/openenv/browsergym_llm.py]: Good point I guess

File: examples/scripts/openenv/catch.py [hunk 7]
```diff
@@ -228,86 +229,112 @@ def main():
...
+        def move(self, direction: str) -> str:
+            """Move the paddle left or right.
+
+            Args:
+                direction: Direction to move, either "left" or "right".
+
+            Returns:
+                The observation after moving.
+            """
+            action_id = 0 if direction == "left" else 2
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/scripts/openenv/catch.py]: Good point as well
